### PR TITLE
Part 2: Creating New Loss Function Classes

### PR DESCRIPTION
**Pull request by emersodb**
# PR Type
Feature

# Short Description

Clickup Ticket(s): [Link](https://app.clickup.com/t/8687xebzz)

This PR is Part 1 of several PRs aiming to simplify and decouple our implementations of FENDA and PerFCL. In this first part, I'm simply creating loss function classes that will be used by the different clients. These functions are stand-alone in this PR to be easier to review.

There are also some small documentation changes and an variable name change for the Ditto and MR-MTL clients.

# Tests Added

Stand-along tests for each of the loss functions implemented, along with one for the pre-existing drift loss function.

### CODE REVIEW

File: fl4health/losses/perfcl_loss.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+                clients. Shape (1, n_samples, n_features)
+        Returns:
+            Tuple[torch.Tensor, torch.Tensor]: Tuple containing the two components of the PerFCL loss function to
+                be weighted and summed.
+        """
+
+        global_feature_loss = self.global_feature_contrastive_loss(
+            features=global_features,  # (z_s)
+            positive_pairs=initial_global_features,  # (z_g)
+            negative_pairs=old_global_features,  # (\hat{z_s})
```
Comment (sanaAyrml) [inline fl4health/losses/perfcl_loss.py:61]: I think I prefer to move initial_global_features.unsqueeze(0) and old_global_features.unsqueeze(0) here rather than doing it outside.

File: fl4health/losses/perfcl_loss.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        )
+        local_feature_loss = self.local_feature_contrastive_loss(
+            features=local_features,  # (z_p)
+            positive_pairs=old_local_features,  # (\hat{z_p})
+            negative_pairs=initial_global_features,  # (z_g)
+        )
```
Comment (sanaAyrml) [inline fl4health/losses/perfcl_loss.py:67]: I think I prefer to move old_local_features.unsqueeze(0) and initial_global_features.unsqueeze(0) here rather than doing it outside.

File: tests/losses/test_contrastive_loss.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    global_features = torch.tensor([[[1.0, 2.0], [3.0, 4.0]], [[5.0, 6.0], [7.0, 8.0]]])
+    local_features = torch.tensor([[[1.0, 2.0], [2.0, 1.0]], [[2.0, 1.0], [1.0, -1.0]]])
+    previous_local_features = torch.tensor([[[1.0, 2.0], [2.0, 1.0]], [[1.0, 1.0], [1.0, 1.0]]])
+
+    loss_value = contrastive_loss(
+        features=local_features.reshape(len(local_features), -1),
+        positive_pairs=global_features.reshape(1, len(global_features), -1),
+        negative_pairs=previous_local_features.reshape(1, len(previous_local_features), -1),
+    )
+
```
Comment (sanaAyrml) [inline tests/losses/test_contrastive_loss.py:22]: I would also prefer to check device of input tensors whether they are actually on same device as loss or not.

File: tests/losses/test_perfcl_loss.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    initial_global_features = torch.tensor([[[1.0, 2.0], [3.0, 4.0]], [[5.0, 6.0], [7.0, 8.0]]])
+    global_features = torch.tensor([[[1.0, 2.0], [2.0, 1.0]], [[2.0, 1.0], [1.0, -1.0]]])
+    old_global_features = torch.tensor([[[1.0, 2.0], [2.0, 1.0]], [[1.0, 1.0], [1.0, 1.0]]])
+    local_features = torch.tensor([[[1.0, 2.0], [3.0, 4.0]], [[5.0, 6.0], [7.0, 8.0]]])
+    old_local_features = torch.tensor([[[1.0, 2.0], [3.0, 4.0]], [[5.0, 6.0], [7.0, 8.0]]])
+
+    perfcl_loss_function = PerFclLoss(DEVICE, global_feature_loss_temperature=0.5, local_feature_loss_temperature=0.5)
+
+    global_feature_loss, _ = perfcl_loss_function(
+        local_features.reshape(len(local_features), -1),
```
Comment (sanaAyrml) [inline tests/losses/test_perfcl_loss.py:40]: same as mentioned above I prefer to move this unsqueeze to inside loss.

**Review by sanaAyrml — COMMENTED** [review]
Overall looks good, just proposed some minor changes.

File: fl4health/losses/contrastive_loss.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    def forward(
+        self, features: torch.Tensor, positive_pairs: torch.Tensor, negative_pairs: torch.Tensor
+    ) -> torch.Tensor:
+        """
+        Compute the contrastive loss based on the features, positive pair and negative pairs. While every feature
+        has a positive pair, it can have multiple negative pairs. The loss is computed based on the similarity
+        between the feature and its positive pair.
+
+        Args:
+            features (torch.Tensor): Main features, shape (n_samples, n_features)
```
[the remark this answers is not in the document]
Comment (emersodb) [reply fl4health/losses/contrastive_loss.py:35]: Will do!

File: fl4health/losses/cosine_similarity_loss.py [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+    def __init__(self, device: torch.device, dim: int = -1) -> None:
+        super().__init__()
+        self.cosine_similarity_function = nn.CosineSimilarity(dim=dim).to(device)
+
+    def forward(self, x1: torch.Tensor, x2: torch.Tensor) -> torch.Tensor:
+        """
+        Assumes that the tensors are provided "batch first"
+
+        Args:
+            x1 (torch.Tensor): First set of tensors to compute cosine sim
```
[the remark this answers is not in the document]
Comment (emersodb) [reply fl4health/losses/cosine_similarity_loss.py:15]: Sounds good

File: fl4health/losses/perfcl_loss.py (hunk 1 above)
Comment (emersodb) [reply fl4health/losses/perfcl_loss.py:61]: Sure, can do.

File: tests/losses/test_contrastive_loss.py (hunk 3 above)
Comment (emersodb) [reply tests/losses/test_contrastive_loss.py:22]: Yeah that's a good call.

File: tests/losses/test_perfcl_loss.py (hunk 4 above)
Comment (emersodb) [reply tests/losses/test_perfcl_loss.py:40]: Sure!

File: tests/losses/test_contrastive_loss.py (hunk 3 above)
Comment (emersodb) [reply tests/losses/test_contrastive_loss.py:22]: Because our github machines are all cpu driven. It's hard to write an "automatic" test for this. However, I logged into an A40 on the cluster and ran the following script to verify that the tensors move properly.
```
import torch

from fl4health.losses.perfcl_loss import PerFclLoss

DEVICE = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")

perfcl_loss_function = PerFclLoss(DEVICE, global_feature_loss_temperature=0.5, local_feature_loss_temperature=0.5)

local_features = torch.tensor([[1, 1, 1], [1, 1, 1]]).float().to("cpu")
global_features = torch.tensor([[1, 1, 1], [1, 1, 1]]).float().to("cpu")
old_local_features = torch.tensor([[0, 0, 0], [0, 0, 0]]).float().to("cpu")
old_global_features = torch.tensor([[0, 0, 0], [0, 0, 0]]).float().to("cpu")
initial_global_features = torch.tensor([[1, 1, 1], [1, 1, 1]]).float().to("cpu")

global_feature_loss, local_feature_loss = perfcl_loss_function(
    local_features,
    old_local_features,
    global_features,
    old_global_features,
    initial_global_features,
)
```
and everything worked well!

**Review by sanaAyrml — APPROVED** [review]

File: fl4health/clients/basic_client.py [hunk 7]
```diff
@@ -893,13 +894,14 @@ def set_optimizer(self, config: Config) -> None:
         self.optimizers = {"global": optimizer}
 
     def clone_and_freeze_model(self, model: nn.Module) -> nn.Module:
-        """Clone and freeze model for use in various loss calculation.
+        """
+        Clones and freezes a models weights. This is to preserve the model in its current state while decoupling in
```
Comment (jewelltaylor) [inline fl4health/clients/basic_client.py:898]: Maybe its just me but the way this is worded is a tad confusing. Perhaps: Creates a clone of the model with frozen weights to be used in loss calculations so the original model is preserved in its current state.

Comment (emersodb) [reply fl4health/clients/basic_client.py:898]: Sounds good!

File: fl4health/losses/contrastive_loss.py [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+        self.cos_sim = torch.nn.CosineSimilarity(dim=-1).to(self.device)
+        self.ce_criterion = torch.nn.CrossEntropyLoss().to(self.device)
+
+    def forward(
+        self, features: torch.Tensor, positive_pairs: torch.Tensor, negative_pairs: torch.Tensor
+    ) -> torch.Tensor:
+        """
+        Compute the contrastive loss based on the features, positive pair and negative pairs. While every feature
+        has a positive pair, it can have multiple negative pairs. The loss is computed based on the similarity
+        between the feature and its positive pair.
```
Comment (jewelltaylor) [inline fl4health/losses/contrastive_loss.py:32]: nit: similarity between the feature and its positive pair relative to negative pairs

File: fl4health/losses/perfcl_loss.py [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+            global_features (torch.Tensor): Features produced by the global feature extractor of the model during the
+                client-side training. Denoted as z_s in the original paper. Shape (batch_size, n_features)
+            old_global_features (torch.Tensor): Features produced by the FINAL global feature extractor of the model
+                from the PREVIOUS server round. Denoted as hat{z}_s in the original paper.
+                Shape (batch_size, n_features)
+            initial_global_features (torch.Tensor): Features produced by the INITIAL global feature extractor of the
+                model at the start of client-side training. This feature extractor is the AGGREGATED weights across
+                clients. Shape (batch_size, n_features)
+        Returns:
+            Tuple[torch.Tensor, torch.Tensor]: Tuple containing the two components of the PerFCL loss function to
```
Comment (jewelltaylor) [inline fl4health/losses/perfcl_loss.py:54]: nit: Maybe including that the first is global and send is local. Will be useful to know if looking at hosted documentation (when we get around to that) and code is not right there

File: research/picai/prepare_data.py [hunk 10]
```diff
@@ -273,24 +273,33 @@ def main() -> None:
     )
 
     args = parser.parse_args()
+
     if args.size is not None and len(args.size) != 3:
         raise ValueError("Argument size must have length 3")
-
-    if args.physical_size is not None and len(args.physical_size) != 3:
-        raise ValueError("Argument physical_size must have length 3")
+    size = (int(args.size(0)), int(args.size(1)), int(args.size(2))) if args.size else None
```
Comment (jewelltaylor) [inline research/picai/prepare_data.py:279]: Thanks for cleaning this up :)

File: fl4health/losses/contrastive_loss.py (hunk 8 above)
Comment (emersodb) [reply fl4health/losses/contrastive_loss.py:32]: 👍

File: fl4health/losses/contrastive_loss.py [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+                "Each feature can have only one positive pair. ",
+                "Thus positive pairs should be a tensor of shape (1, batch_size, n_features) ",
+                f"rather than {positive_pairs.shape}",
+            )
+        positive_pair = positive_pairs[0]
+        assert len(features) == len(positive_pair)
+        positive_similarity = self.cos_sim(features, positive_pair)
+        logits = positive_similarity.reshape(-1, 1)
+        for negative_pair in negative_pairs:
+            assert len(features) == len(negative_pair)
```
[the remark this answers is not in the document]
Comment (emersodb) [reply fl4health/losses/contrastive_loss.py:59]: I don't think there is anything stopping us from doing that. Will implement.

File: fl4health/losses/perfcl_loss.py (hunk 9 above)
Comment (emersodb) [reply fl4health/losses/perfcl_loss.py:54]: Yeah good point.

File: research/picai/prepare_data.py (hunk 10 above)
Comment (emersodb) [reply research/picai/prepare_data.py:279]: It was throwing a mypy error for me. Not really sure why it didn't throw on your PR but it's all good 🙂

File: fl4health/losses/contrastive_loss.py [hunk 12]
```diff
@@ -51,15 +70,19 @@ def forward(
...
+        # Compute similarity of the batch of features with the provided batch of positive pair features
+        positive_similarity = self.cosine_similarity_function(features, positive_pair)
+        # Store similarities with shape batch_size x 1
         logits = positive_similarity.reshape(-1, 1)
-        for negative_pair in negative_pairs:
-            assert len(features) == len(negative_pair)
-            negative_similarity = self.cos_sim(features, negative_pair)
-            logits = torch.cat((logits, negative_similarity.reshape(-1, 1)), dim=1)
+
+        # Compute the similarity of the batch of features with the collection of batches of negative pair features
```
Comment (jewelltaylor) [inline fl4health/losses/contrastive_loss.py:81]: Nice!

**Review by jewelltaylor — APPROVED** [review]
Thanks for the changes, looks great!
